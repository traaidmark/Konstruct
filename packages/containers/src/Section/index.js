// CONTAINERS [ SECTION ] ######################################################

// 1. DEPENDENCIES +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

// 1.1. EXTERNAL DEPENDENCIES ..................................................

import React, { useCallback } from 'react';
import propTypes from 'prop-types';

// 1.1. END ....................................................................

// 1.2. INTERNAL DEPENDENCIES ..................................................

import Styled from './stylesheet';

// 1.2. END ....................................................................

// 1. END ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

// 2. COMPONENT ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

const Section  = ({
  children,
  className,
  style,
  noTop,
  noBottom,
  noLeft,
  noRight,
  variantX = 'default',
  variantY = 'default',
  scrollX = false,
  scrollY = false,
  heading,
  description,
  footerHeading,
  footerActions,
}) => {

  // 2.1. FUNCTIONS ............................................................

  // 2.2.1. RENDER HEADER

  const Header = useCallback(() => {
    if (heading || description) {
      return(
        <Styled.header>
          { heading && <h2>{ heading }</h2> }
          { description && <p>{ description }</p> }
        </Styled.header>
      );
    };
    return null;
  }, [heading, description]);

  // 2.2.1. END

  // 2.2.2. RENDER FOOTER

  const Footer = useCallback(() => {
    if (footerHeading || footerActions) {
      return(
        <Styled.footer>
          { footerHeading && <h3>{ footerHeading }</h3> }
          { footerActions && footerActions }
        </Styled.footer>
      );
    };
    return null;
  }, [footerHeading, footerActions]);


  // 2.2.2. END

  // 2.1. END ..................................................................

  // 2.2. RENDER COMPONENT .....................................................

  return (

    <Styled.container
      style={ style }
      className={ className }
      noTop={ noTop }
      noBottom={ noBottom }
      noLeft={ noLeft }
      noRight={ noRight }
      variantX={ variantX }
      variantY={ variantY }
      scrollX={ scrollX }
      scrollY={ scrollY }
    >
      <Header />
      { children }
      <Footer />
    </Styled.container>

  );

  // 2.2. END ..................................................................

};

// 2. END ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


// 3. PROP-TYPES +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Section.propTypes = {
  children: propTypes.node.isRequired,
  style: propTypes.object,
  className: propTypes.string,
  heading: propTypes.oneOfType([
    propTypes.string,
    propTypes.node
  ]),
  description: propTypes.oneOfType([
    propTypes.string,
    propTypes.node
  ]),
  footerHeading: propTypes.oneOfType([
    propTypes.string,
    propTypes.node
  ]),
  footerActions: propTypes.array,
  noTop: propTypes.bool,
  noBottom: propTypes.bool,
  noLeft: propTypes.bool,
  noRight: propTypes.bool,
  variantX: propTypes.oneOf([
    'default',
    'wide',
    'narrow',
  ]),
  variantY: propTypes.oneOf([
    'default',
    'shallow',
    'deep',
  ]),
  scrollX: propTypes.bool,
  scrollY: propTypes.bool,
};

// 3. END ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

export default Section;

// END OF FILE #################################################################
