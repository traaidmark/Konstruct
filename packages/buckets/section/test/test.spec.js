// BUCKET [ SECTION > TESTS ] ##################################################

// 1. DEPENDENCIES +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

// 1.1. EXTERNAL DEPENDENCIES ..................................................

import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

// 1.1. END ....................................................................

// 1.2. INTERNAL DEPENDENCIES ..................................................

import Props from './props';

import Section from '../src';

// 1.2. END ....................................................................

// 1. END ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

// 2. TEST +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

// 2.1. RENDER WITHOUT PROPS ...................................................

describe('Bucket > Section', () => {
  test('renders correctly without any props', () => {
    const tree = renderer.create(
      <Section>{ 'Test' }</Section>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// 2.1. END ....................................................................

// 2.2. RENDER WITH ALL PROPS ..................................................

describe('Bucket > Section', () => {
  test('renders correctly with all props', () => {
    const tree = renderer.create(
      <Section {...Props }>
        { 'Test' }
      </Section>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// 2.2. END ....................................................................

// 2. END ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

// END OF FILE #################################################################
