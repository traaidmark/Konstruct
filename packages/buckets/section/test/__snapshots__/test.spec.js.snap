// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bucket > Section renders correctly with all props 1`] = `
<section
  className="b-section b-section--shallow"
>
  <header
    className="b-section__header"
  >
    <h2>
      Section title
    </h2>
    <p>
      Nulla leo odio, ullamcorper vel enim sed, maximus pretium sapien. Duis a augue pharetra ex 
      <strong>
        consequat ultrices.
      </strong>
    </p>
  </header>
  Test
  <footer
    className="b-section__footer"
  >
    <h3>
      Sign up today!
    </h3>
    <ul
      className="u-list-horizontal u-list-horizontal--center"
    >
      <li>
        <a>
          Link 1
        </a>
      </li>
      <li>
        <a>
          Link 2
        </a>
      </li>
    </ul>
  </footer>
</section>
`;

exports[`Bucket > Section renders correctly without any props 1`] = `
<section
  className="b-section "
>
  Test
</section>
`;
